### PR TITLE
add option to lazily precompute multiplication tables

### PR DESCRIPTION
**Pull request by oberstet**
this is an attempt to fix https://github.com/warner/python-ecdsa/issues/180

the does not change any existing behavior, but adds new behavior:

when `PYTHON_ECDSA_ENABLE_LAZY_PRECOMPUTE` env var is set, the multiplication tables for curves are only precomputed upon first use.

without the env var being set (the current behavior), tables for _all_ supported curves are precomputed at import time regardless of whether the curves are used in the user app at all. 

effect can be seen in this test:

```
(cpy382_2) oberstet@intel-nuci7:~/scm/3rdparty/python-ecdsa$ python test.py 
python-ecdsa v0.16.0+1.g3388afc
0.4552568449998944
(cpy382_2) oberstet@intel-nuci7:~/scm/3rdparty/python-ecdsa$ PYTHON_ECDSA_ENABLE_LAZY_PRECOMPUTE=1 python test.py 
python-ecdsa v0.16.0+1.g3388afc
0.020854599999438506
(cpy382_2) oberstet@intel-nuci7:~/scm/3rdparty/python-ecdsa$ 
```

using `test.py`:

```python
import timeit

def test1():
    from ecdsa.curves import SECP256k1

print(timeit.timeit(test1, number=1))
```

### CODE REVIEW

File: src/ecdsa/__init__.py
Comment (tomato42) [inline src/ecdsa/__init__.py:47]: yes, it's wrong but it's here for backwards compatibility

File: src/ecdsa/ellipticcurve.py
Comment (tomato42) [inline src/ecdsa/ellipticcurve.py:188]: this is not thread safe, if two threads use the same object it will end up with malformed `__precompute` table; see `scale()`

File: src/ecdsa/ellipticcurve.py
Comment (tomato42) [inline src/ecdsa/ellipticcurve.py:184]: I don't like the environment variable idea, this only makes the use of the library more complex, and delayed precomputation benefits everybody that doesn't use all curves every time

**Review by tomato42 — CHANGES_REQUESTED** [review]

**Comment by tomato42** [comment]
The CI also failed: https://travis-ci.org/github/warner/python-ecdsa/builds/742220130

File: src/ecdsa/ellipticcurve.py
Comment (oberstet) [reply src/ecdsa/ellipticcurve.py:188]: as far as I see: all call contexts of `maybe_precompute` already have critical sections

Comment (oberstet) [reply src/ecdsa/ellipticcurve.py:188]: and the original pre-computation in the ctor doesn't need one .. I think? an object cannot be accessed from 2 threads while it is still being constructed?

File: src/ecdsa/__init__.py
Comment (oberstet) [reply src/ecdsa/__init__.py:47]: ok, I'll readd it

File: src/ecdsa/ellipticcurve.py
Comment (oberstet) [reply src/ecdsa/ellipticcurve.py:184]: note sure I understand: so should I remove the controlling flag altogether and make it "always lazy"?

Comment (tomato42) [reply src/ecdsa/ellipticcurve.py:184]: yes, make it always lazy

File: src/ecdsa/ellipticcurve.py
Comment (tomato42) [reply src/ecdsa/ellipticcurve.py:189]: technically, if you mess around `__new__` and mess with inheritance really badyly, you could end up with two threads in an `__init__` at the same time, but not in regular use, so yes, we didn't need locks before because in practice it's not possible to have two threads in `__init__` at the same time

> as far as I see: all call contexts of maybe_precompute already have critical sections

huh? as long as two threads start execution of the _maybe_precompute at the same time, and one thread checks the if before the second thread runs the first `.append()` both threads will execute the entirety of it

and even if not, then the other thread can think the __precompute list is fully initialised when it's in the process of being initialised, ending up using incomplete table

File: src/ecdsa/ellipticcurve.py
Comment (tomato42) [inline src/ecdsa/ellipticcurve.py:232]: it's not needed here

File: src/ecdsa/ellipticcurve.py
Comment (tomato42) [inline src/ecdsa/ellipticcurve.py:221]: it's not needed here

File: src/ecdsa/ellipticcurve.py
Comment (tomato42) [inline src/ecdsa/ellipticcurve.py:593]: other is an integer here, not a point

**Comment by oberstet** [comment]
ok, pushed adjustments based on your feedback, and tested on cpy and pypy:

* https://gist.github.com/oberstet/905251d53a5168ea8c56996618a3a4cb
* https://gist.github.com/oberstet/c8ed75f1be47f2eacbfccc9b25d5565d

2 notes:

* in the original code, the precomputation was called with constructor x/y whereas in the new code, x/y can be a gmp wrapped x/y. I tested both the non-gmp and gmp tox envs, so that should be good?
* I now added a writer (single I guess) lock within the precompute block. new locks, means new risks for deadlock. also that lock will now happen later on .. lazy (which is the whole point).

File: src/ecdsa/ellipticcurve.py
Comment (oberstet) [inline src/ecdsa/ellipticcurve.py:587]: this will escalate the lock from reader to writer! not sure ... is "getstate" supposed to return also the lazy-computed attribute?

Comment (tomato42) [reply src/ecdsa/ellipticcurve.py:587]: > this will escalate the lock from reader to writer! not sure ...

I don't see how...

> is "getstate" supposed to return also the lazy-computed attribute?

yes, it should; the pickling is used for multiprocess module for example

File: src/ecdsa/ellipticcurve.py
Comment (tomato42) [inline src/ecdsa/ellipticcurve.py:186]: and what if multiple threads execute this line at the same time?

I'm also not a fan of overloading the `_scale_lock`, it's used for a different purpose

Comment (oberstet) [reply src/ecdsa/ellipticcurve.py:186]: > and what if multiple threads execute this line at the same time?

it will block in https://docs.python.org/3.8/library/threading.html#threading.Lock.acquire

a deadlock would arise if that code blocking would have previously acquired a write lock upon which the precomputing thread would have to wait subsequently. 

anyways, we'll use our fork for now, works

Comment (tomato42) [reply src/ecdsa/ellipticcurve.py:186]: what I mean is that both threads will then add values to the precompute list, that's wrong

Comment (oberstet) [reply src/ecdsa/ellipticcurve.py:186]: no, because of `not self.__precompute` on line 182. or are you saying that this latter access has to be under the critical section as well

Comment (oberstet) [reply src/ecdsa/ellipticcurve.py:186]: .. but that would escalate each and every "maybe_precompute" call, even when already cached. another option would be to optain a read lock already for the initial access in "maybe_precompute", and then escalate that to write lock without releasing before. not sure if the lock wrapper in use even allows that. mmh. locks;) don't use them ..

**Comment by tomato42** [comment]
> ok, pushed adjustments based on your feedback, and tested on cpy and pypy:
> 
>     * https://gist.github.com/oberstet/905251d53a5168ea8c56996618a3a4cb
> 
>     * https://gist.github.com/oberstet/c8ed75f1be47f2eacbfccc9b25d5565d
> 
> 
> 2 notes:
> 
>     * in the original code, the precomputation was called with constructor x/y whereas in the new code, x/y can be a gmp wrapped x/y. I tested both the non-gmp and gmp tox envs, so that should be good?

check the CI results: https://travis-ci.org/github/warner/python-ecdsa/pull_requests
 
>     * I now added a writer (single I guess) lock within the precompute block. new locks, means new risks for deadlock. also that lock will now happen later on .. lazy (which is the whole point).

the solution to deadlocking is not locking in fewer places but to lock in correct places; here we need to acquire reader lock when checking if the precompute is initialised, free it, and then acquire a writer lock if it wasn't initialised; then after getting the writer lock we need to check again for initialisation

**Comment by oberstet** [comment]
> here we need to acquire reader lock when checking if the precompute is initialised, free it, and then acquire a writer lock if it wasn't initialised; then after getting the writer lock we need to check again for initialisation

I've updated the PR which should reflect ^ now ..

File: src/ecdsa/ellipticcurve.py
Comment (tomato42) [inline src/ecdsa/ellipticcurve.py:37]: unused import

File: src/ecdsa/ellipticcurve.py
Comment (tomato42) [inline src/ecdsa/ellipticcurve.py:186]: it's `_scale_lock` because it was used for `scale()` but now we also use it for precompute... maybe change its name to `_update_lock`?

**Review by tomato42 — CHANGES_REQUESTED** [review]

**Comment by tomato42** [comment]
also few previous versions failed in CI because of excessive line length, you may want to run `tox -e codechecks` before pushing a new version

**Comment by oberstet** [comment]
thanks for your time and for reviewing! pls merge.

> sidenote: the CI settings on travis for this repo don't allow for unbuilt pushes to the PR supersede others (to CI run requests pile up), and for whatever reasons, a single CI run takes hours.

**Comment by tomato42** [comment]
the builds take hours because travis limited the number of free workers recently: https://www.traviscistatus.com/#month don't know why exactly, unfortunately I don't have access to administer it to fix the dismissal of stale PRs or lack of CI status in PRs either... I'll send an email to Warner tomorrow

**Review by tomato42 — APPROVED** [review]
ok, looks good, but I'll wait until the CI finishes before merging

**Comment by oberstet** [comment]
thanks for approving!

> re: travis: ah, right. interesting - news for me. guess we'll run into this soonish as well on our oss stuff ..

**Comment by tomato42** [comment]
merged, thank you!

**Pull request merged**
